programs/vote/src/vote_state/mod.rs

```diff
@@ -1885,12 +1983,349 @@ mod tests {
+    #[allow(clippy::field_reassign_with_default)]
+    fn test_timely_vote_credits() {
+        // Each element of test data is:
+        // (vote_state_votes: Vec<Slot>,
+        //  next_vote_slots: Vec<Slot>,
+        //  voted_in_slot: Slot,
+        //  expected_credits: u64)
+        let test_data: Vec<(Vec<Slot>, Vec<Slot>, Slot, u64)> = vec![
+            // No slots finalized yet
+            (vec![0, 1, 2], vec![4], 5, 0),
+            // 1 slot finalized with latency 0
+            (
+                vec![
+                    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
+                    22, 23, 24, 25, 26, 27, 27, 29, 30,
+                ],
+                vec![31],
+                32,
+                16,
+            ),
+            // 2 slots finalized with latency 1 and 0
+            (
+                vec![
+                    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
+                    22, 23, 24, 25, 26, 27, 27, 29, 30,
+                ],
+                vec![31, 32],
+                33,
+                32,
+            ),
+            // 3 slots finalized with latency 2, 1 and 0
+            (
+                vec![
+                    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
+                    22, 23, 24, 25, 26, 27, 27, 29, 30,
+                ],
+                vec![31, 32, 33],
+                34,
+                48,
+            ),
+            // 4 slots finalized with latency 3, 2, 1 and 0
+            (
+                vec![
+                    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
+                    22, 23, 24, 25, 26, 27, 27, 29, 30,
+                ],
+                vec![31, 32, 33, 34],
+                35,
+                64,
+            ),
+            // 5 slots finalized with latency 4, 3, 2, 1 and 0
+            (
+                vec![
+                    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
+                    22, 23, 24, 25, 26, 27, 27, 29, 30,
+                ],
+                vec![31, 32, 33, 34, 35],
+                36,
+                79,
+            ),
+            // 5 slots finalized with latency 6, 5, 4, 3, 2
+            (
+                vec![
+                    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
+                    22, 23, 24, 25, 26, 27, 27, 29, 30,
+                ],
+                vec![31, 32, 33, 34, 35],
+                38,
+                74,
+            ),
+            // 1 slot finalized with latency 200
+            (
+                vec![
+                    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
+                    22, 23, 24, 25, 26, 27, 27, 29, 30,
+                ],
+                vec![31],
+                232,
+                1,
+            ),
+        ];
+
```

Comment:
lets add some tests where the voting is not consecutive and slots are getting popped off in the tower

Reply:
I tried to do so with the last one but I will do a few more.